omigami_client/spec2vec.py

```diff
@@ -37,18 +39,35 @@ def match_spectra_from_path(self, mgf_path: str, n_best: int) -> List[pd.DataFra
         A list of pandas dataframes containing the best matches.
 
         """
+
+        # gets generator
         spectra_generator = load_from_mgf(mgf_path)
-        payload = self._build_payload(spectra_generator, n_best)
-        api_request = self._send_request(payload)
-        prediction = self._format_results(api_request)
 
-        return prediction
+        # issue requests respecting the spectra limit per request
+        predictions = []
+        while True:
+            payload, done = self._build_payload(spectra_generator, n_best)
+            if not done:
+                api_request = self._send_request(payload)
+                predictions.extend(self._format_results(api_request))
+            else:
+                break
+        return predictions
 
     def _build_payload(
         self, spectra_generator: Generator[Spectrum, None, None], n_best_spectra: int
-    ) -> Payload:
-        """Extract abundance pairs and Precursor_MZ data, then build the json payload"""
+    ) -> (JSON, bool):
```

Comment:
why are we returning a tuple here? The signature looks very weird by building a payload a returning a tuple with the payload and the boolean.

I think it is cleaner if we build the payload in batches, so it returns a `List[Payload]` then we just loop through them:

```python
payloads = self._build_payloads(spectra_generator, n_best)
predictions = []

for payload in payloads:
    api_request = self._send_request(payload)
    predictions.extend(self._format_results(api_request))
```

Reply:
hm yeah, I liked your suggestion, but thinking more about it, I think your way requires the user to hold all payloads in memory at once (you have a list with all of it in the outer scope) while the way it is is just the payload for the request at hand (once the loop iteration sends the request, the last payload gets overwritten). And since spec2vec uses a generator to feed only the data necessary, I think it would be best for large payloads if it were like this (I tried copying the payload generated from the first integration test, which isn't even the largest one, and my Pycharm ran out of memory, it warned and crashed :P)

Reply:
I changed the loop a bit see if you like it more  x ]

Reply:
I still don't like it because of the while loop and the method with the weird signature/interface. I understand the memory issue. Then I think we should loop through the generator:

```python
batch = []
for spectrum in spectra_generator:
    batch.append(spectrum)

    if len(batch) == BATCH_SIZE:
        payload = self.build_payload(batch)
        api_request = self._send_request(payload)
        predictions.extend(self._format_results(api_request))
        batch = []
```

Reply:
This way of batching is actually very common in machine learning for training neural networks where your data is too big to be held in memory

Reply:
that code snippet won't work if the num of spectra is not divisible by BATCH_SIZE, but I can take some inspiration from it, thanks.

Reply:
Tried making it look cleaner. Can't escape the while loop though, otherwise would need to repeat the build_payload/send_request/format_results block, which I think is worse